Pegasus/include/geometry.hpp

```diff
@@ -0,0 +1,875 @@
+        Vector3 sphereMassCenter;
+        double sphereRadius;
+        double penetration;
+    };
+
+    template <>
+    struct Cache<Plane, Box> : CacheBase
+    {
+        Vector3 boxMassCenter;
+        std::array<Vector3, 3> boxAxes;
+        std::array<Vector3, 8> boxVertices;
+        std::array<Vector3, 6> boxFaces;
+        std::array<double, 6> boxFaceDistances;
+        std::array<double, 8> boxPenetrations;
+        Vector3 planeNormal;
+        double planeDistance;
+    };
+
+    template <>
+    struct Cache<Sphere, Plane> : CacheBase
+    {
+        Cache<Plane, Sphere> psCache;
+    };
+
+    template <>
+    struct Cache<Sphere, Sphere> : CacheBase
+    {
+        Vector3 bMassCenter;
+        Vector3 baVector;
+        double bRadius;
+        Vector3 aMassCenter;
+        double aRadius;
+        double radiusSum;
+    };
+
+    template <>
+    struct Cache<Sphere, Box> : CacheBase
+    {
+        Vector3 boxMassCenter;
+        std::array<Vector3, 3> boxAxes;
+        std::array<Vector3, 6> boxNormals;
+        std::array<Vector3, 6> boxFaces;
+        std::array<double, 6> boxFaceDistances;
+        std::array<Vector3, 8> boxVertices;
+        std::array<double, 8> boxVerticesProjections;
+        std::array<Vector3, 4> separatingAxes;
+        Vector3 sphereMassCenter;
+        double sphereRadius;
+    };
+
+    template <>
+    struct Cache<Box, Plane> : CacheBase
+    {
+        Cache<Plane, Box> pbCache;
+        Vector3 planeMassCenter;
+        Vector3 boxContactNormal;
+    };
+
+    template <>
+    struct Cache<Box, Sphere> : CacheBase
+    {
+        Cache<Sphere, Box> sbCache;
+    };
+
+    template <>
+    struct Cache<Box, Box> : CacheBase
+    {
+        Vector3 aMassCenter;
+        Vector3 bMassCenter;
+        std::array<Vector3, 8> aBoxVertices, bBoxVertices;
+        std::array<Vector3, 6> aBoxAxes, bBoxAxes;
+        std::array<Vector3, 6> aBoxFaces, bBoxFaces;
+        std::vector<Vector3> separatingAxes;
+        std::array<double, 6> aBoxFaceDistances, bBoxFaceDistances;
+        std::array<double, 8> aBoxVerticesDistances, bBoxVerticesDistances;
+        Vector3 contactNormal;
+        double penetration = 0;
+    };
+
+    // General intersection queries
+    template < typename ShapeA, typename ShapeB >
+    void initialize(SimpleShape const * a, SimpleShape const * b, CacheBase  * cacheBase);
+
+    template < typename ShapeA, typename ShapeB >
+    bool overlap(SimpleShape const * a, SimpleShape const * b, CacheBase  * cacheBase);
+
+    template < typename ShapeA, typename ShapeB >
+    Vector3 calculateContactNormal(SimpleShape const * a, SimpleShape const * b, CacheBase  * cacheBase);
+
+    template < typename ShapeA, typename ShapeB >
+    double calculatePenetration(SimpleShape const * a, SimpleShape const * b, CacheBase  * cacheBase);
+
+    // Plane, Plane
+    template <>
+    inline void initialize<Plane, Plane>(SimpleShape const *a, SimpleShape const *b, CacheBase *cacheBase)
```

Comment:
It maybe or maybe not. It's called quite frequently, so it's a tradeoff. Only profiling could tell for sure.